### PR TITLE
Infinite horizontal scrolling in 2D

### PR DESCRIPTION
**Pull request by bagnell**
Adds infinite scrolling in 2D.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Merge in master.

**Comment by pjcozzi** [comment]
The right click zoom out starts to translate strangely at max zoom out.  Can we clamp it?

**Comment by pjcozzi** [comment]
What happens if the viewer sets the camera height above the new max in 2D?

**Comment by pjcozzi** [comment]
In 2D, use the geocoder to fly to `-180, 0`.  It flies to the wrong location and crashes here and in master.

**Comment by pjcozzi** [comment]
How painful do you think it is to fix this lighting issue?

ok

```javascript
var greenCircle = viewer.entities.add({
    position: Cesium.Cartesian3.fromDegrees(-179.0, 40.0),
    name : 'Green circle at height',
    ellipse : {
        semiMinorAxis : 1500000.0,
        semiMajorAxis : 1500000.0,
        height: 0.0,
        material : Cesium.Color.GREEN.withAlpha(0.5)
    }
});
```

**Comment by pjcozzi** [comment]
Can you test all the geometry types across the IDL?  If they require an easy fix, let's do it.  If not, submit an issue.  For example, polyline volume does not work:
```javascript
var redTube = viewer.entities.add({
    name : 'Red tube with rounded corners',
    polylineVolume : {
        positions : Cesium.Cartesian3.fromDegreesArray([-190.0, 32.0,
                                                        -85.0, 36.0,
                                                        -89.0, 36.0]),
        shape : computeCircle(60000.0),
        material : Cesium.Color.RED
    }
});
```

image

**Comment by pjcozzi** [comment]
Actually, polyline volume is OK...we should test longitude in -180 to 360...might need some doc updates...doesn't need to all be in this PR, but let's at least scope it.
```json
var redCorridor = viewer.entities.add({
    name : 'Red corridor on surface with rounded corners and outline',
    corridor : {
        positions : Cesium.Cartesian3.fromDegreesArray([
                                                        100.0, 40.0,
                                                        200.0, 40.0,
                                                    ]),
                                                    width : 200000.0,
        material : Cesium.Color.RED.withAlpha(0.5),
        outline : true,
        outlineColor : Cesium.Color.RED
    }
});
```

**Comment by pjcozzi** [comment]
What is the plan for polylines?  Fix them now or write an issue?  Seems like this will be one of the most common use cases for infinite 2D.
image

```javascript
var redLine = viewer.entities.add({
    name : 'Red line on the surface',
    polyline : {
        positions : Cesium.Cartesian3.fromDegreesArray([75, 35,
                                                        200, 35]),
        width : 5,
        material : Cesium.Color.RED
    }
});
```

**Comment by pjcozzi** [comment]
There are JSHint failures: https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/111261539

File: Source/Scene/Scene.js [hunk 1]
```diff
@@ -1643,76 +1649,162 @@ define([
...
+
+            Camera.clone(savedCamera, camera);
+        } else {
+            viewport.x = 0;
+            viewport.y = 0;
+            viewport.width = context.drawingBufferWidth;
+            viewport.height = context.drawingBufferHeight;
+
+            if (mode !== SceneMode.SCENE2D) {
+                updatePrimitives(scene);
```
Comment (pjcozzi) [inline Source/Scene/Scene.js]: This series of function calls is duplicated several times below.  Should we put them in a separate function?

**Comment by pjcozzi** [comment]
Does this need any new tests?

**Comment by pjcozzi** [comment]
That's all my comments.  The changes to scene are fine for now, but we'll need to improve the overall structure when we expose multiple viewports.

**Comment by bagnell** [comment]
>In 2D, use the geocoder to fly to -180, 0. It flies to the wrong location and crashes here and in master.

This is broken in master too and doesn't just crash at +/-180. I think the geocoder is flying to the lon/lat in Earth fixed.

**Comment by pjcozzi** [comment]
>> In 2D, use the geocoder to fly to -180, 0. It flies to the wrong location and crashes here and in master.

> This is broken in master too and doesn't just crash at +/-180. I think the geocoder is flying to the lon/lat in Earth fixed.

Can you submit an issue or fix if it is easy?

**Comment by bagnell** [comment]
>What happens if the viewer sets the camera height above the new max in 2D?

The camera position and orientation are regular Javascript properties so there is nothing stopping someone from setting the height above the max or rotaing in 2D. We could put checks when the camera is updated, but that can be overwritten on preRender. The best way would be make them ES5 properties, but the camera is used pretty heavily so it might have a significant performance impact.

CC @mramato @shunter

**Comment by bagnell** [comment]
Actually, most code that updates the camera does something like:

`Cartesian3.clone(newPosition, camera.position);`

so ES5 properties won't work. We could probably do the camera checks right before render, like below the max height in 2D, orientation vectors are orthonormal, etc.

File: Source/Scene/Scene.js (hunk 1 above)
Comment (bagnell) [inline Source/Scene/Scene.js]: I thought that too, but they need to be executed in this order and I didn't want to execute anything that I didn't have to twice.

**Comment by pjcozzi** [comment]
> We could probably do the camera checks right before render, like below the max height in 2D, orientation vectors are orthonormal, etc.

OK with me.  Also just documenting it is OK with me.

File: Source/Scene/Scene.js (hunk 1 above)
Comment (pjcozzi) [inline Source/Scene/Scene.js]: I don't get it; can't we just make a function that calls these functions, and then call it from here and the three places below?

**Comment by bagnell** [comment]
>Can you test all the geometry types across the IDL?

Everything seems to work except for polylines and they all have the same lighting problem.

**Comment by pjcozzi** [comment]
Let me know when this is ready.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look. I'll open issues for the lighting and the polyline issues. I'm looking into them now.

**Comment by bagnell** [comment]
>How painful do you think it is to fix this lighting issue?

This issue is fixed. It wasn't a lighting issue; it was drawing primitives in the second viewport twice. Now I'm looking into the polyline issue.

**Comment by bagnell** [comment]
@pjcozzi The polyline issues have been fixed in master. This is ready for another review.

**Comment by pjcozzi** [comment]
Retesting https://github.com/AnalyticalGraphicsInc/cesium/pull/3621#issuecomment-187791831.  Do you get this artifact?  We can live with it if it is not trivial to fix.
ok

**Comment by pjcozzi** [comment]
Retesting https://github.com/AnalyticalGraphicsInc/cesium/pull/3621#issuecomment-187805511.  Do you get these artifacts?  Again, we can live with them if need be.
ok

**Comment by pjcozzi** [comment]
Do you get this artifact in 2D when zooming in and out in the G&A example?  It is not in master.

ok

**Comment by pjcozzi** [comment]
Testing with the GeoJSON Sandcastle example, the selection indicator turns off early when scrolling to the right:
ok

**Comment by pjcozzi** [comment]
For https://github.com/AnalyticalGraphicsInc/cesium/pull/3621#issuecomment-199372489, it might be easier to reproduce it in the GeoJSON Sandcastle example:

ok

**Comment by pjcozzi** [comment]
For https://github.com/AnalyticalGraphicsInc/cesium/pull/3621#issuecomment-199371917, check out the Geoeye orbit in the CZML Sandcastle example.  The line left of the IDL (or in the left viewport?) jitters.

ok

Also go fullscreen, and you'll see several vertical black line artifacts on the left side.

**Comment by bagnell** [comment]
For https://github.com/AnalyticalGraphicsInc/cesium/pull/3621#issuecomment-199372489, it also happens in master after clicking the cardboard button. Its most likely an issue with converting to window coordinates in the shader when there are multiple viewports. I'll open a PR into master when its fixed.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another review.

**Comment by pjcozzi** [comment]
Do you get the black flicker on the edge of the viewport when panning in the CZML Sandcastle example:

ok

**Comment by pjcozzi** [comment]
Only one AGI billboard is highlighted here.  Is this just the way it is for now?

image

**Comment by pjcozzi** [comment]
What do you suggest for labels (and billboards I imagine) crossing the IDL/viewport?  Write up an issue with known 2D issues?

Using the Sandcastle Picking example:

ok

**Comment by pjcozzi** [comment]
Also if you mouse over above or below in the Picking Sandcastle example, the label moves to the wrong place.

**Comment by pjcozzi** [comment]
Just those comments.

**Comment by bagnell** [comment]
>Do you get the black flicker on the edge of the viewport when panning in the CZML Sandcastle example?

No, i haven't seen that. The IDL should only visible at most in one location. How did you zoom out that far? Your AGI billboard comment is the same issue.

**Comment by bagnell** [comment]
>What do you suggest for labels (and billboards I imagine) crossing the IDL/viewport? Write up an issue with known 2D issues?

Yes, I'm not sure how to fix that without computing the corner points of the billboard, projecting them to 2D and splitting them across the IDL each frame.

**Comment by pjcozzi** [comment]
Tested on Mac.

All looks good!

**Pull request merged**